main.go

```diff
@@ -42,6 +52,77 @@ type Output struct {
 	Expiration      string
 }
 
+type Cache struct {
+	Dir string
+}
+
+func NewCache(dir string) (*Cache, error) {
+	if _, err := os.Open(dir); err != nil {
+		return nil, err
+	}
+
+	fn := path.Join(dir, "validate")
+	if _, err := os.Create(fn); err != nil {
+		return nil, err
+	}
+	os.Remove(fn)
+
+	return &Cache{
+		Dir: dir,
+	}, nil
+}
+
+func (c *Cache) filenameByRole(role string) string {
+	h := sha256.New()
+	h.Write([]byte(role))
+	return fmt.Sprintf("%x", h.Sum(nil))
+}
+
+func (c *Cache) Get(role string) (*Output, bool) {
+	fp := path.Join(c.Dir, c.filenameByRole(role))
+	if f, err := os.Open(fp); err != nil {
+		return nil, false
+	} else {
+
+		if bytes, err := io.ReadAll(f); err != nil {
+			return nil, false
+		} else {
+			output := Output{}
+			if err := json.Unmarshal(bytes, &output); err != nil {
+				return nil, false
+			}
+
+			// check expiration
+			if tm, err := time.Parse(time.RFC3339, output.Expiration); err == nil && tm.After(time.Now().Add(REFRESH_HEAD_ROOM)) {
+				return &output, true
+			}
+
+			// don't use cache
+			return nil, false
```

Comment:
Suggestion to get rid of elses and test all errors

```suggestion
	f, err := os.Open(fp)
	if err != nil {
		return nil, false
	}

	bytes, err := io.ReadAll(f)
	if err != nil {
		return nil, false
	}

	var output Output
	if err := json.Unmarshal(bytes, &output); err != nil {
		return nil, false
	}

	tm, err := time.Parse(time.RFC3339, output.Expiration)
	if err != nil {
		return nil, false
	} 

	// check expiration
	if !tm.After(time.Now().Add(REFRESH_HEAD_ROOM)) {
		return nil, false
	}

	return &output, false
```